File: lru_cache.py
class LRUCache: # 3 seconds, faster than only 1% solutions

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = []

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        # print("get", key, self.cache)
        i = len(self.cache)
        value = None
        for k in range(len(self.cache)):
            if key in self.cache[k]:
                i = k
                value = self.cache[k][key]
                break
        
        if i < len(self.cache): # found
            del self.cache[i]
            self.cache.append({key:value})
            return self.cache[-1][key]
        else:
            return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        # print("put", key, value, self.cache)
        i = len(self.cache)
        for k in range(len(self.cache)):
            if key in self.cache[k]:
                del self.cache[k]
                self.cache.append({key:value})
                i = k
                break
        
        if i == len(self.cache): # unsuccessful search
            if len(self.cache) < self.capacity: 
                self.cache.append({key:value})
            else:
                del self.cache[0]
                self.cache.append({key:value})


class LRUCache: # 576ms, faster than only 13% solutions

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = {}
        self.recent = 0

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        # print("get", key, self.cache)
        if key in self.cache:
            self.recent += 1
            self.cache[key][1] = self.recent
            return self.cache[key][0]
        else:
            return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        # print("put", key, value, self.cache)
        if key in self.cache:
            self.cache[key][0] = value
        else:
            if len(self.cache) == self.capacity:
                # remove the least recently used cache
                mini = self.recent+1
                minkey = mini
                for i in self.cache.keys():
                    if self.cache[i][1] < mini:
                        minkey = i
                        mini = self.cache[i][1]
                del self.cache[minkey]
            self.cache[key] = [value,-1]
            
        self.recent = (1 + self.recent)
        self.cache[key][1] = self.recent
        

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)


class LRUCache(dict): # 96ms solution, credits: LeetCode

    def __init__(self, capacity):
        self.capacity = capacity

    def get(self, key):
        if key not in self:
            return -1
        
        val = self.pop(key)
        self[key] = val
        return val

    def put(self, key, value):
        if key in self:
            self.pop(key)
            
        if len(self) == self.capacity:
            self.pop(next(iter(self)))
            
        self[key] = value

File: lru_cache.cpp
class LRUCache { // beats 13.51% of solutions, 112 ms and 10.8 MB
public:
    unordered_map<int, int> d;
    vector<int> queue;
    int capacity;
    
    LRUCache(int cap) {
        capacity = cap;
    }
    
    int get(int key) {
        if(d.find(key) == d.end()){
            return -1;
        }
        int val = d[key];
        queue.erase(remove(queue.begin(), queue.end(), key), queue.end()); // Erase-remove idiom
        queue.push_back(key);
        // cout<<"get"<<" "<<key<<endl;
        // for(auto i:queue){
        //     cout<<i<<endl;
        // }
        // cout<<" dict below:";
        // for(auto i:d){
        //     cout<<i.first<<":"<<i.second<<endl;
        // }
        // cout<<endl;
        return val;
    }
    
    void put(int key, int value) {
        if(d.find(key) != d.end()){
            d.erase(key);
            queue.erase(remove(queue.begin(), queue.end(), key), queue.end());
        }
        if(d.size() == capacity){
            // cout<<"hit capacity! del: "<<*queue.begin()<<endl;
            d.erase(*queue.begin());
            queue.erase(queue.begin());
        }
        d.insert(make_pair(key,value));
        queue.push_back(key);
        // cout<<"put"<<" "<<key<<" "<<value<<endl;
        // for(auto i:queue){
        //     cout<<i<<endl;
        // }
        // cout<<" dict below:";
        // for(auto i:d){
        //     cout<<i.first<<":"<<i.second<<endl;
        // }
        // cout<<endl;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

// could have used doubly linked list

// class LRUCache { // Credits - Leetcode
// public:
//     LRUCache(int capacity) : capacity_(capacity), head_(nullptr), tail_(nullptr) {}

//     int get(int key) {
//         // key not exists
//         auto it = hash_.find(key);
//         if (it == hash_.end())
//             return -1;

//         // key exists
//         // 1. get the node
//         // 2. move the node to the head
//         int ret = it->second->value;
//         move2head(it->second);
//         return ret;
//     }

//     void put(int key, int value) {
//         // key exists
//         auto it = hash_.find(key);
//         if (it != hash_.end()) {
//             Node* node = it->second;
//             node->value = value;
//             move2head(node);
//             return;
//         }

//         // key not exists
//         // 1. enough capacify
//         if (hash_.size() < capacity_) {
//             Node* node = new Node(key, value);
//             hash_[key] = node;
//             move2head(node);
//             return;
//         }
//         // 2. not enough capacity
//         Node* node = tail_;
//         removetail();
//         node->key = key;
//         node->value = value;
//         move2head(node);
//         hash_[key] = node;
//     }

// private:
//     struct Node {
//         int key;
//         int value;
//         Node* prev;
//         Node* next;

//         Node(): prev(nullptr), next(nullptr) {}
//         Node(int k, int v): key(k), value(v), prev(nullptr), next(nullptr) {}
//     };

//     // 1. tail_ cannot be nullptr
//     // 2. remove it from the linkedlist
//     // 3. remove it from hash table
//     void removetail() {
//         hash_.erase(tail_->key);
//         if (tail_->prev == nullptr) {
//             head_ = nullptr;
//             tail_ = nullptr;
//         } else {
//             Node* prev = tail_->prev;
//             prev->next = nullptr;
//             tail_->prev = nullptr;
//             tail_ = prev;
//         }
//     }

//     // 1. delete this node from linked list
//     //    there's a case the node is not in the linked list
//     // 2. add it into head
//     void move2head(Node* node) {
//         if (head_ == nullptr) {
//             head_ = tail_ = node;
//             return;
//         }
//         if (node == head_)
//             return;
//         if (node == tail_) {
//             node->prev->next = nullptr;
//             tail_ = node->prev;
//             node->prev = nullptr;
//             node->next = head_;
//             head_->prev = node;
//             head_ = node;
//             return;
//         }
//         if (node->prev == nullptr && node->next == nullptr) {
//             head_->prev = node;
//             node->next = head_;
//             head_ = node;
//             return;
//         }
//         node->prev->next = node->next;
//         node->next->prev = node->prev;
//         node->prev = nullptr;
//         node->next = head_;
//         head_->prev = node;
//         head_ = node;
//     }

//     int capacity_;
//     Node* head_;
//     Node* tail_;    // delete from the tail
//     unordered_map<int, Node*> hash_;
// };